### DISCUSSION TITLE
Queries from Grafana

### DISCUSSION THREAD

**Discussion by DrRuhe** [obi1kenobi/trustfall#273] [Ideas]
I thought it could be nice to allow trustfall queries to be executed from Grafana. So, I jumped into the rabbit hole and drafted a simple architecture that could support that. 

```mermaid
flowchart LR

    subgraph Grafana[Grafana Process]
        UI -- Utilizes --> Plugin[Grafana-Trustfall Plugin]
    end
    Plugin -- sends queries --> Webserver

    subgraph Server
        Webserver -- executes queries --> Trustfall
    end
```

Basically two things are required to enable this use case. A Grafana-Plugin that enables trustfall Datasources and a way to expose trustfall adapters via HTTP.

- The Grafana plugin would connect to the server and send the queries from Grafana. The plugin could also provide a fancy query editor, but that's for a later version.  In theory, it is also possible to bundle the query engine in the Grafana-Plugin using wasm, but this would require users to replace the plugins every time the adapter changes.

- A standardized template webserver that is generic over the exposed adapter could be a nice fit for exposing Adapters over HTTP (possibly implement [HttpServiceFactory for an Adapter, so an adapter can be easily exposed as a HTTP route via actix-web](https://docs.rs/actix-web/latest/actix_web/struct.App.html#method.service)).

I might implement this myself, depending on the free time available. 
I wanted to put it out there, so others can provide further ideas and so that I don't forget these thoughts.

**Comment by obi1kenobi** [comment]
This looks awesome! I'm not familiar with Grafana plugins, but if there's anything I can do on the Trustfall end to make this easier for you to build, I'm all ears!

**Reply by DrRuhe** [reply]
I am a bit confused about what the best approach to executing queries is. I see that trustfall::execute_query exists. But then in the HYTRADBOI example, you build a custom execute_query function. Should I build a custom execute_query function?

I am currently facing lifetime/type mismatch problems as the trustfall::execute_query function expects parameters in one way but actix-web provides them in another. So I tried making trustfall::execute_query more generic. In general, I am seeing a lot of Arc<_> usage in the code. Could you elaborate on why those are needed? In particular for BTreeMaps of arguments. I understand that the keys might be reusable, so the Arc<str> there make sense. Do we need them to reduce the number of lifetimes on things like InterpretedQuery?

**Reply by obi1kenobi** [reply]
The HYTRADBOI example code predates the `trustfall` crate, and I must have forgotten to update it. PRs welcome!

Please do use the `trustfall::execute_query` function instead of building your own.

Just in case, might you be open to pair-programming some of this, say tomorrow early evening your time? I've never tried to plug in actix-web into Trustfall, and I'd be interested in seeing the friction points so I can improve Trustfall's ergonomics and write better "getting started" docs.

The `Arc` use is indeed to reduce the number of lifetimes. Trustfall has only one lifetime: the lifetime of the data the `Adapter` sits on top of, and everything else is refcounted to simplify the implementation. The glue code between actix-web and Trustfall won't be hot from a performance perspective, so a bit of cloning to simplify the lifetimes situation shouldn't be noticeable. But like I mentioned, I'm interested in seeing more about the specific issues in case there's a better approach I can guide users toward.

**Reply by DrRuhe** [reply]
Sure, I'd be open to pair programming this.  I have time to work on this now (for the next 4-5h), so just send me a meeting invite to my mail address.

**Reply by DrRuhe** [reply]
A few more Questions have popped up:

- Is there an Adapter somewhere I can use easily? I was about to use the rustdoc adapters, but those have to be initialized somehow.
- Is it possible to run Queries that utilize Multiple adapters and if so, how does this work? Is there something like `From<AdapterA::Vertex> for AdapterB::Vertex` that would allow queries to cross adapter boundaries?

In general, the current API is difficult to work with, as it requires specifics that are not easily fulfilled in the context of actix-web. For example, acquiring an `Rc<Adapter>` for a request is difficult, as I only get access to an `Arc<Adapter>` but I cannot easily create an Rc without introducing a `Clone` bound (which e.g. the RustdocAdapter does not implement). I think a new layer of abstraction in the trustfall crate could solve this. Possibly an `Engine` struct, where Adapters can be registered and the queries can be executed via something like: `Engine::query(&self,query:AsRef<str>,variables: impl QueryArguments)->Result<...>`. But I feel like I am missing the required details, so I'd like your Feedback on that first.

**Reply by obi1kenobi** [reply]
I sent you a meeting invite to your email address!

**Reply by obi1kenobi** [reply]
You can make a rustdoc file by running `cargo rustdoc -- -w json -Zunstable-options` in a lib crate, and then grabbing the rustdoc JSON file from inside the `target/doc` directory.

I am working on making it possible to run cross-adapter queries! I have a prototype running locally but the API is hard to get right and not everything is implemented yet, so it'll be a bit longer before I can publish that.

Great to know about the `Arc` vs `Rc` on `Adapter` being problematic. In a future version I might make `Arc<Adapter>` the default, or make it possible to support both. I think there's a way to generalize over `Arc` and `Rc` as well, but I need to look up the blog post that showed how to do that.

In the meantime, an unfortunate but probably working hack might be to make an `Arc<Rc<Adapter>>` 😅

**Reply by DrRuhe** [reply]
After our session I got stuck on actix-web not wanting to parse my requests. After being busy with other stuff for a few days, I looked at it again today. I got it to work under axum and now I can send queries to a webserver and it responds with the answers. With that, I also noticed, that trustfall::execute_query is a library function that returns anyhow::Result. It should define distinct error variants (perhaps utilizing thiserror) to allow proper error handling.

As the Grafana Integration only really makes sense once queries can utilize multiple adapters I will wait with the implementation until that's possible.

**Reply by obi1kenobi** [reply]
`trustfall::execute_query` is just a façade in front of the underlying `trustfall_core` APIs: https://github.com/obi1kenobi/trustfall/blob/main/trustfall/src/lib.rs#L69-L87

Many changes to `thiserror`-like error types are breaking changes, and I didn't want every incremental improvement there to be a breaking change. This is why the `trustfall` façade layer erases the error type info. The underlying `trustfall_core` implementations use `thiserror` error types like you suggest.

If you're okay with somewhat more common breaking changes, replace `trustfall::execute_query` with analogous direct use of the underlying `trustfall_core` functionality.

> As the Grafana Integration only really makes sense once queries can utilize multiple adapters I will wait with the implementation until that's possible.

Can you give me an example of "queries utilizing multiple adapters" here? I feel like could mean one of several things, some of which are already possible today and I'd hate to keep you waiting if I don't have to :)

**Reply by obi1kenobi** [reply]
I'm about to merge PR #286 which makes query execution use `Arc<Adapter>` instead of `Rc<Adapter>`. Hopefully that will simplify this use case! I'd love feedback on whether it worked, or if there's something else that is needed too.

It's a major breaking change so if I didn't fully fix the problem, I'd love to group as many of the breaking fixes I need into one release so as not to annoy users needlessly with too many consecutive breaking changes.

**Reply by obi1kenobi** [reply]
I just released Trustfall v0.5.0 with the `Arc<Adapter>` change! Thanks for pointing out the problem that needed to be fixed there.
